paddle/fluid/platform/mkldnn_reuse.h

```diff
@@ -436,6 +436,158 @@ class ActivationMKLDNNHandler : public MKLDNNHandler {
   std::shared_ptr<mkldnn::eltwise_backward::primitive_desc> activation_bwd_pd_;
 };
 
+class LRNMKLDNNHandler : public MKLDNNHandler {
+ public:
+  LRNMKLDNNHandler(bool is_test, const platform::MKLDNNDeviceContext& dev_ctx,
+                   mkldnn::engine engine, const std::string& base_key)
+      : platform::MKLDNNHandler(dev_ctx, engine, base_key), is_test_(is_test) {}
+
+  std::shared_ptr<mkldnn::lrn_forward::primitive_desc>
+  AcquireLRNPrimitiveDescriptor(const mkldnn::memory::desc& src_md, const int n,
+                                const float alpha, const float beta,
+                                const float k) {
+    // LRN PD has to be passed to Grad op that
+    // may be executed by diffrent thread, hence
+    // for that one we use key that does not contain TID
+    const std::string key_lrn_pd = key_common_ + "@lrn_pd";
+    fwd_pd_ = std::static_pointer_cast<mkldnn::lrn_forward::primitive_desc>(
+        dev_ctx_.GetBlob(key_lrn_pd));
+    if (fwd_pd_ == nullptr) {
+      static std::mutex acquire_barrier;
+      std::lock_guard<std::mutex> block_threads_until_finish_this_job(
+          acquire_barrier);
+      fwd_pd_ = std::static_pointer_cast<mkldnn::lrn_forward::primitive_desc>(
+          dev_ctx_.GetBlob(key_lrn_pd));
+      if (fwd_pd_ == nullptr) {
+        auto forward_desc = mkldnn::lrn_forward::desc{
+            is_test_ ? mkldnn::prop_kind::forward_inference
+                     : mkldnn::prop_kind::forward_training,
+            mkldnn::lrn_across_channels, src_md, n, alpha, beta, k};
+        fwd_pd_.reset(
+            new mkldnn::lrn_forward::primitive_desc(forward_desc, engine_));
+        dev_ctx_.SetBlob(key_lrn_pd, fwd_pd_);
+      }
+    }
+    return fwd_pd_;
+  }
+
+  std::shared_ptr<mkldnn::memory> AcquireWorkspaceMemory(void) {
+    // workspace has to be passed to Grad op that
+    // may be executed by diffrent thread, hence
+    // for that one we use key that does not contain TID
+    auto local_key = key_common_ + "@workspace";
+    const std::string key_lrn_pd = key_common_ + "@lrn_pd";
+    fwd_pd_ = std::static_pointer_cast<mkldnn::lrn_forward::primitive_desc>(
+        dev_ctx_.GetBlob(key_lrn_pd));
+    // PD from FWD op has to exist.
+    PADDLE_ENFORCE(fwd_pd_ != nullptr, "LRN PD MKL-DNN not found in cache!");
```

Comment:
Please consider moving the fetching of workspace primitive desc into the if condition of memory primitive.

Reply:
ok. done